File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseAdapter.java
package br.com.jonathanzanella.myexpenses.expense;

import android.content.Intent;
import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.support.annotation.WorkerThread;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.joda.time.DateTime;

import java.util.List;

import br.com.jonathanzanella.myexpenses.MyApplication;
import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.bill.Bill;
import br.com.jonathanzanella.myexpenses.chargeable.Chargeable;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.helpers.AdapterColorHelper;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import butterknife.BindView;
import butterknife.ButterKnife;

class ExpenseAdapter extends RecyclerView.Adapter<ExpenseAdapter.ViewHolder> {
	private final ExpenseRepository expenseRepository;
	private final ExpenseAdapterPresenter presenter;
	private List<Expense> expenses;
	private DateTime date;

	public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		@BindView(R.id.row_expense_name)
		TextView name;
		@BindView(R.id.row_expense_date)
		TextView date;
		@BindView(R.id.row_expense_value)
		TextView value;
		@BindView(R.id.row_expense_chargeable)
		TextView chargeable;
		@BindView(R.id.row_expense_bill_stt)
		TextView billViewStt;
		@BindView(R.id.row_expense_bill)
		TextView billView;
		@BindView(R.id.row_expense_charge_next_month)
		TextView chargeNextMonth;

		private final AdapterColorHelper adapterColorHelper;

		public ViewHolder(View itemView) {
			super(itemView);

			ButterKnife.bind(this, itemView);

			//noinspection deprecation
			int oddColor = itemView.getContext().getResources().getColor(R.color.color_list_odd);
			//noinspection deprecation
			int evenColor = itemView.getContext().getResources().getColor(R.color.color_list_even);
			adapterColorHelper = new AdapterColorHelper(oddColor, evenColor);

			itemView.setOnClickListener(this);
		}

		@UiThread
		public void setData(final Expense expense) {
			itemView.setTag(expense.getUuid());
			itemView.setBackgroundColor(adapterColorHelper.getColorForGridWithTwoColumns(getAdapterPosition()));
			name.setText(expense.getName());
			date.setText(Expense.SIMPLE_DATE_FORMAT.format(expense.getDate().toDate()));
			value.setText(CurrencyHelper.format(expense.getValue()));
			new AsyncTask<Void, Void, Chargeable>() {

				@Override
				protected Chargeable doInBackground(Void... voids) {
					return expense.getChargeableFromCache();
				}

				@Override
				protected void onPostExecute(Chargeable c) {
					super.onPostExecute(c);
					chargeable.setText(c.getName());
				}
			}.execute();

			chargeNextMonth.setVisibility(expense.isChargedNextMonth() ? View.VISIBLE : View.INVISIBLE);
			new AsyncTask<Void, Void, Bill>() {

				@Override
				protected Bill doInBackground(Void... voids) {
					return expense.getBill();
				}

				@Override
				protected void onPostExecute(Bill bill) {
					super.onPostExecute(bill);
					if(bill == null) {
						billViewStt.setVisibility(View.INVISIBLE);
						billView.setVisibility(View.INVISIBLE);
					} else {
						billViewStt.setVisibility(View.VISIBLE);
						billView.setVisibility(View.VISIBLE);
						billView.setText(bill.getName());
					}
				}
			}.execute();
		}

		@Override
		public void onClick(View v) {
			Expense expense = getExpense(getAdapterPosition());
			if(expense != null) {
                Intent i = new Intent(itemView.getContext(), ShowExpenseActivity.class);
                i.putExtra(ShowExpenseActivity.KEY_EXPENSE_UUID, expense.getUuid());
                itemView.getContext().startActivity(i);
			}
		}
	}

	ExpenseAdapter() {
		expenseRepository = new ExpenseRepository(new RepositoryImpl<Expense>(MyApplication.getContext()));
		presenter = new ExpenseAdapterPresenter(this, expenseRepository);
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_expense, parent, false);
		return new ViewHolder(v);
	}

	@Override
	public void onBindViewHolder(ViewHolder holder, int position) {
		holder.setData(getExpense(position));
	}

	@Override
	public int getItemCount() {
		return expenses != null ? expenses.size() : 0;
	}

	@WorkerThread
	public void loadData(DateTime date) {
		expenses = expenseRepository.monthly(date);
		expenses = presenter.getExpenses(true, date);
		this.date = date;
	}

	@Nullable
	private Expense getExpense(int position) {
		return expenses != null ? expenses.get(position) : null;
	}

	public void filter(String filter) {
		presenter.filter(filter);
		expenses = presenter.getExpenses(false, date);
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseAdapter.kt
package br.com.jonathanzanella.myexpenses.expense

import android.content.Intent
import android.os.AsyncTask
import android.support.annotation.UiThread
import android.support.annotation.WorkerThread
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import br.com.jonathanzanella.myexpenses.MyApplication
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.bill.Bill
import br.com.jonathanzanella.myexpenses.chargeable.Chargeable
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl
import br.com.jonathanzanella.myexpenses.helpers.AdapterColorHelper
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper
import br.com.jonathanzanella.myexpenses.transaction.Transaction
import kotlinx.android.synthetic.main.row_expense.view.*
import org.joda.time.DateTime

internal open class ExpenseAdapter : RecyclerView.Adapter<ExpenseAdapter.ViewHolder>() {
    private val expenseRepository = ExpenseRepository(RepositoryImpl<Expense>(MyApplication.getContext()))
    private val presenter: ExpenseAdapterPresenter
    private var expenses: List<Expense> = ArrayList()
    private var date: DateTime? = null

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val adapterColorHelper: AdapterColorHelper

        init {
            val oddColor = ResourcesCompat.getColor(itemView.context.resources, R.color.color_list_odd, null)
            val evenColor = ResourcesCompat.getColor(itemView.context.resources, R.color.color_list_even, null)
            adapterColorHelper = AdapterColorHelper(oddColor, evenColor)

            itemView.setOnClickListener(this)
        }

        @UiThread
        fun setData(expense: Expense) {
            itemView.tag = expense.uuid
            itemView.setBackgroundColor(adapterColorHelper.getColorForGridWithTwoColumns(adapterPosition))
            itemView.name.text = expense.name
            itemView.date.text = Transaction.SIMPLE_DATE_FORMAT.format(expense.getDate().toDate())
            itemView.value.text = CurrencyHelper.format(expense.value)
            object : AsyncTask<Void, Void, Chargeable>() {

                override fun doInBackground(vararg voids: Void): Chargeable? {
                    return expense.chargeableFromCache
                }

                override fun onPostExecute(c: Chargeable?) {
                    super.onPostExecute(c)
                    itemView.chargeable.text = c?.name
                }
            }.execute()

            itemView.chargeNextMonth.visibility = if (expense.isChargedNextMonth) View.VISIBLE else View.INVISIBLE
            object : AsyncTask<Void, Void, Bill>() {

                override fun doInBackground(vararg voids: Void): Bill? {
                    return expense.bill
                }

                override fun onPostExecute(bill: Bill?) {
                    super.onPostExecute(bill)
                    if (bill == null) {
                        itemView.billStt.visibility = View.INVISIBLE
                        itemView.bill.visibility = View.INVISIBLE
                    } else {
                        itemView.billStt.visibility = View.VISIBLE
                        itemView.bill.visibility = View.VISIBLE
                        itemView.bill.text = bill.name
                    }
                }
            }.execute()
        }

        override fun onClick(v: View) {
            val expense = getExpense(adapterPosition)
            val i = Intent(itemView.context, ShowExpenseActivity::class.java)
            i.putExtra(ShowExpenseActivity.KEY_EXPENSE_UUID, expense.uuid)
            itemView.context.startActivity(i)
        }
    }

    init {
        presenter = ExpenseAdapterPresenter(expenseRepository)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.row_expense, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setData(getExpense(position))
    }

    override fun getItemCount(): Int {
        return expenses.size
    }

    @WorkerThread
    fun loadData(date: DateTime) {
        expenses = expenseRepository.monthly(date) as ArrayList<Expense>
        expenses = presenter.getExpenses(true, date)
        this.date = date
    }

    private fun getExpense(position: Int): Expense {
        return expenses[position]
    }

    fun filter(filter: String) {
        presenter.filter(filter)
        expenses = presenter.getExpenses(false, date!!)
    }
}